### PR TITLE
Vue 3 Compatibility

### PR DESCRIPTION
**Pull request by subzero10**
## Status
**READY**

## Description
- Update package to support vue.js v3

## Related PRs
honeybadger-io/docs#vue3 | [Vue.js v3 Support](https://github.com/honeybadger-io/docs/pull/137)

## Todos
- [x] npm update (vue.js 3 + deps)
- [x] npm audit (outdated deps)
- [x] Vue.js v3 compatibility
- [x] Honeybadger.notify to correctly send `context` object - ~~`parentVNodeTag`~~ `parentName`
- [x] Unit Tests
- [x] Prep for release to @honeybadger-io/vue@next
- [x] Documentation
- [x] Changelog Entry (unreleased)

## Steps to Test or Reproduce
Outline the steps to test or reproduce the PR here.
```bash
> git pull --prune
> git checkout <branch>
> npm i
```

To test:
```bash
> npm test
```

To run:
```bash
> HONEYBADGER_API_KEY=YOUR_API_KEY npm run dev
```

### CODE REVIEW

File: build/webpack.base.conf.js [hunk 1]
```diff
@@ -36,7 +36,7 @@ module.exports = {
   resolve: {
     extensions: [".js", ".vue", ".json"],
     alias: {
-      vue$: "vue/dist/vue.esm.js",
+      //vue$: "vue/dist/vue.esm.js",
```
Comment (subzero10) [inline build/webpack.base.conf.js:39]: I couldn't find why this was here.

File: build/webpack.base.conf.js [hunk 2]
```diff
@@ -84,16 +84,6 @@ module.exports = {
     ]
   },
   node: {
-    // prevent webpack from injecting useless setImmediate polyfill because Vue
-    // source contains it (although only uses it if it's native).
-    setImmediate: false,
```
Comment (subzero10) [inline build/webpack.base.conf.js:89]: Not valid options anymore.

File: build/webpack.dev.conf.js [hunk 3]
```diff
@@ -58,13 +45,19 @@ const devWebpackConfig = merge(baseWebpackConfig, {
...
     }),
     // copy custom static assets
-    new CopyWebpackPlugin([
-      {
-        from: path.resolve(__dirname, '../static'),
-        to: config.dev.assetsSubDirectory,
-        ignore: ['.*']
-      }
-    ])
+    new CopyWebpackPlugin({
```
Comment (subzero10) [inline build/webpack.dev.conf.js:48]: New syntax.

File: build/webpack.dev.conf.js [hunk 4]
```diff
@@ -23,27 +23,14 @@ const devWebpackConfig = merge(baseWebpackConfig, {
...
-    open: config.dev.autoOpenBrowser,
-    overlay: config.dev.errorOverlay
-      ? { warnings: false, errors: true }
-      : false,
-    publicPath: config.dev.assetsPublicPath,
-    proxy: config.dev.proxyTable,
-    quiet: true, // necessary for FriendlyErrorsPlugin
-    watchOptions: {
-      poll: config.dev.poll,
-    }
```
Comment (subzero10) [inline build/webpack.dev.conf.js:46]: Not valid options anymore.

File: config/index.js [hunk 5]
```diff
@@ -33,7 +33,7 @@ module.exports = {
      */
 
     // https://webpack.js.org/configuration/devtool/#development
-    devtool: 'cheap-module-eval-source-map',
+    devtool: 'eval-cheap-module-source-map',
```
Comment (subzero10) [inline config/index.js:36]: Renamed.

File: demo/components/Miniwolf.vue [hunk 6]
```diff
@@ -29,13 +29,13 @@ export default {
       console.log(this.something)
     }
   },
-  filters: {
```
Comment (subzero10) [inline demo/components/Miniwolf.vue:32]: Filters not supported anymore.

File: demo/components/Miniwolf.vue [hunk 7]
```diff
@@ -3,7 +3,7 @@
     <h1>{{ msg }}</h1>
     <button id="componentErrantButton" v-on:click="makeSomethingUnrenderable()">Trigger a component Error</button>
     <button id="setValueButton" v-on:click="setSomethingValue(100)">Set a value (no error)</button>
-    <div>{{something | formatSomething}}</div>
```
Comment (subzero10) [inline demo/components/Miniwolf.vue:6]: `|` not supported anymore.

File: build/webpack.base.conf.js (hunk 1 above)
Comment (joshuap) [reply build/webpack.base.conf.js:39]: Hrmm, not sure. I'm OK w/ deleting if it works without it.

**Review by joshuap — COMMENTED** [review]
@subzero10 this is looking good! I tested it out locally, and unit tests pass. `npm run e2e` fails, however, I don't remember running the end-to-end tests before, so I'm not sure of the last time they were passing. I also tried the sample app (`npm run dev`) and it worked as expected.

What is left on this PR?

Also, I'd like your opinion on the best approach for supporting both vue 2 and vue 3 in our NPM package. On GitHub, the `master` branch supports vue 2, and the `vue3` branch support vue 3. We release the `master` branch (vue 2) to NPM, and `vue3` is not currently released to NPM. It looks like the Vue project itself uses the `next` tag on NPM for their vue 3 release. Should we do the same, and if so, can you prep this branch for release to @honeybadger-io/vue@next?

File: build/webpack.base.conf.js (hunk 1 above)
Comment (JasonTrue) [reply build/webpack.base.conf.js:39]: It was probably boilerplate from some sort of create-vue-app analog from the original proof of concept; it was in the initial commit as far as I can tell. It's been too long ago for me to remember why it was in there :P

**Comment by subzero10** [comment]
@joshuap I tried running the end-to-end tests on the master branch, but they fail also. Should I spend some time working on getting them fixed? Or perhaps a better alternative would be to do that in a separate PR, after we merge this?

Regarding on what's left:
- End-to-End tests. Waiting on your feedback for this.
- I don't know what `parentVNodeTag` is, so I will have to try the vue-2 branch and make sure I match the value in vue-3.
- Update documentation and create a changelog entry.

I agree with the `next` tag approach. I will do that!

**Comment by JasonTrue** [comment]
On my really ancient clone, last updated in spring 2019, e2e doesn't run successfully either. It was never part of CI and the one file in `test/e2e/specs/test.js` has only a default test in it that at first glance looks like it's meant to work on a default vue application, not a library. So I think you can dismiss any errors from those tests. E2E tests would probably be a good idea, I suppose? But not originally scoped. It should be safe to remove the reference to e2e in package.json and the `test/e2e` directory, but this needn't be part of this PR.

I can't do much more than provide historical context (which I've largely forgotten since it's been a couple years since I touched this).

parentVNodeTag, if I remember right, basically makes sure the parent component tag (name) is reachable from the Honeybadger Vue instance. My vague memory of why you might need that is that Vue 2 allowed (but didn't necessarily encourage) mounting multiple Vue applications on a page, in which case you might like to know which instance of vue triggered an error if the VM isn't the root element. The tag (name) is more useful for logging than the node itself. But I could be remembering this entirely wrong (I've spent almost no time in Vue in the last 18 months except for some trivial things).

**Comment by subzero10** [comment]
@JasonTrue Thanks for the insight. I too tried going back in time to run e2e. I wasn't sure whether it was my local setup or it was broken 🤷 
 
OK thanks, so `parentVNodeTag` should be the parent node's tag value.

**Comment by subzero10** [comment]
@JasonTrue @joshuap I have some updates re: `parentVnodeTag`:
For context, I tried triggering an error in Honeybadger using **vue-2**:
vue-2-error
The component that threw the error is `Miniwolf`. The value of `parentVnodeTag` is `vue-component-8-Miniwolf`. In vue-2, this is how we got that value:
Screenshot 2021-09-01 at 9 20 41 AM

I am not able to get the same or a similar value from **vue-3**:
Screenshot 2021-09-01 at 9 27 17 AM

I am thinking if it would be OK to replace `parentVnodeTag` with a property named `parentName`, which will have the name of the parent component:
Screenshot 2021-09-01 at 9 27 43 AM

What do you think?

**Comment by subzero10** [comment]
@joshuap Another couple of things:
- I set the package version to `2.0.0`. I figured we should bump the major version for `vue-3` support.
- When you said to prep the branch for `@honeybadger-io/vue@next`, did you mean to set the version in `package.json`? Because I don't have permissions to do anything else (i.e. set dist-tag with npm).

**Comment by joshuap** [comment]
> @joshuap I tried running the end-to-end tests on the master branch, but they fail also. Should I spend some time working on getting them fixed? Or perhaps a better alternative would be to do that in a separate PR, after we release this?
> 
> Regarding on what's left:
> 
> * End-to-End tests. Waiting on your feedback.

I see you've been working with @JasonTrue on the e2e stuff (thank you for the input Jason!!) I'm cool with doing whatever can be done here, but also if you feel like a separate PR would be better, that's totally fine. The e2e tests don't need to run to merge this imo.

Btw, we do have the unit tests running in GitHub Actions afaik, and I'm not sure why they aren't running against this PR. 🤔

> * Documentation. Not sure how are we going to do that, since docs appear to be a different project.

Find the docs [here](https://github.com/honeybadger-io/docs/blob/master/source/lib/javascript/integration/vue.html.md.erb) (you should now have access).

> * I don't know what `parentVNodeTag` is, so I will have to try the vue-2 branch and make sure I match the value in vue-3.

Yeah, I'm not really up to speed on this either, but I see you're making some progress. I'll post a follow-up comment about that.

> * Create a changelog entry.

🙏

> I agree with the `next` tag approach. I will do that!

🎉

> @JasonTrue @joshuap I have some updates re: `parentVnodeTag`:
> For context, I tried triggering an error in Honeybadger using **vue-2**:
> vue-2-error
> The component that threw the error is `Miniwolf`. The value of `parentVnodeTag` is `vue-component-8-Miniwolf`. In vue-2, we got that value from `vm.$options._parentVNode.tag`:
> Screenshot 2021-09-01 at 9 20 41 AM
> 
> I am not able to get the same or a similar value from **vue-3**:
> Screenshot 2021-09-01 at 9 27 17 AM
> 
> I am thinking if it would be OK to replace `parentVnodeTag` with a property named `parentName`, which will have the name of the parent component:
> Screenshot 2021-09-01 at 9 27 43 AM
> 
> What do you think?

Yep, that works for me.



> @joshuap Another couple of things:
> 
> * I set the package version to `2.0.0`. I figured we should bump the major version for `vue-3` support.

How does the Vue project handle versions between the `latest` and `next` tags on NPM? Is `latest` 2.0 and `next` 3.0? If so, what if we bump our `next` tag to version `3.0.0` to sync w/ Vue's major version moving forward?

> * When you said to prep the branch for `@honeybadger-io/vue@next`, did you mean to set the version in `package.json`? Because I don't have permissions to do anything else (i.e. set dist-tag with npm).

Yeah, just set the version in package.json. It would also help me out if you could make a note about how to release the `vue3` branch to `next` in the [releasing section of the readme](https://github.com/honeybadger-io/honeybadger-vue#releasing).

**Comment by subzero10** [comment]
> How does the Vue project handle versions between the latest and next tags on NPM? Is latest 2.0 and next 3.0? If so, what if we bump our next tag to version 3.0.0 to sync w/ Vue's major version moving forward?

Good idea! Just committed a package.json with version 3.0.0.

>Yeah, just set the version in package.json. It would also help me out if you could make a note about how to release the vue3 branch to next in the releasing section of the readme.

Done. In summary:
```
// checkout vue3 branch
npm version 3.0.0
npm publish --tag next
```
To verify, check Versions tab in npmjs.com [here](https://www.npmjs.com/package/@honeybadger-io/vue).

File: build/webpack.base.conf.js (hunk 1 above)
Comment (subzero10) [reply build/webpack.base.conf.js:39]: Thank you both for your input. I will let this commented out since everything seems to work without it.

**Review by joshuap — APPROVED** [review]
This looks good to me, thanks @subzero10!

Go ahead and merge if you think it's ready, then I'll do the release.

**Comment by subzero10** [comment]
@joshuap I merged into `honeybadger-io:vue3` and the CI failed (typescript definitions). Let me look into that first before you release.

**Comment by joshuap** [comment]
> @joshuap I merged into `honeybadger-io:vue3` and the CI failed (typescript definitions). Let me look into that first before you release.

Sure, just let me know when it's ready to go.

**Comment by subzero10** [comment]
@joshuap Typescripts definitions are fixed and the CI works! 🥳 

I committed on the honeybadger-vue#vue3 branch, not on my fork: https://github.com/honeybadger-io/honeybadger-vue/commit/63b9aabaacf87be18fdcfd135a2c99f7c34ca112

Feel free to release this. I will have the docs PR updated asap.

**Pull request merged**